Feature/blocks transactions accounts

jcnelson opened the pull request:
This is just a draft for now, since I'm still working on tests (and will likely be fixing bugs related to that).  But I just wanted to make this available early in case you all wanted to track progress :)   Will update this note with a more detailed code explanation once it's ready for the "real" review.

jcnelson commented:
This PR is not ready for merge, but it is ready for review.  I don't expect there to be any significant revisions to the design or implementation at this point.  However, it needs a lot more end-to-end test coverage before I'll be happy with it.

There's a lot of code here, so I'll be writing an accompanying set of diagrams and explainers for various subsystems in the coming weeks.  This PR adds:

* Block and transaction wire formats
* Block and transaction serialization and deserialization
* Transaction signing and verifying
* Microblock signing and verifying
* Block assembly
* Miner rewards and reward distribution

Still to come:

* Better test coverage on assembling chains of complex blocks and microblocks
* A functional "miner trust" smart contract for accumulating and doling out miner rewards.  The main thrust of this PR is to ensure that the rewards are correctly calculated and granted to a smart contract that can administer their distribution.
* Better test coverage for different reward windows and maturity periods.

To get an understanding of how it works, my advice is to look at the code in the following order:

0. Read SIP 005 in the sip/sip-005 branch, all the way up to this section here: https://github.com/blockstack/blockstack-core/blob/sip/sip-005/sip/sip-005-blocks-and-transactions.md#calculating-a-fully-qualified-object-name.  Read everything above that to understand what this PR does.
1. Read the src/chainstate/stacks/miner.rs code, and its (simplistic) tests.  This describes at a high-level how transactions are assembled into blocks and microblocks.
2. Read the src/chainstate/stacks/mod.rs code, the src/chainstate/stacks/transaction.rs code, and the src/chainstate/stacks/auth.rs code to understand, respectively, (1) the transaction wire formats, (2) the high-level transaction methods and signing/verifying routines, and (3) the nitty-gritty details of how each signature is calculated and verified.
3. Read src/chainstate/stacks/blocks.rs to understand how blocks are structured
4. Read src/chainstate/stacks/db/transactions.rs to understand how transactions are validated within a block, and applied against the Clarity DB state.
5. Read src/chainstate/stacks/db/blocks.rs and src/chainstate/stacks/db/headers.rs to understand how blocks and microblocks and their headers are stored, queued up for processing, and selected and appended to the Stacks chain state.  This is the core of the blockchain validation logic.
6. Read src/chainstate/stacks/db/accounts.rs to understand how miner payouts are tracked, and how the block reward is calculated.
7. Read src/chainstate/stacks/db/mod.rs to understand how the Stacks chain state is instantiated and booted up.

lgalabru commented:
Thank you for providing some guidance @jcnelson!

kantai commented on src/vm/contexts.rs:
```diff
@@ -136,13 +170,26 @@ impl AssetMap {
 
+        for (principal, stx_amount) in other.stx_map.drain() {
+            let next_amount = self.get_next_stx_amount(&principal, stx_amount)?;
+            stx_to_add.push((principal.clone(), next_amount));
```
this clone shouldn't be necessary

kantai commented on src/vm/contexts.rs:
```diff
@@ -136,13 +170,26 @@ impl AssetMap {
+
+        for (principal, stx_burn_amount) in other.burn_map.drain() {
+            let next_amount = self.get_next_stx_burn_amount(&principal, stx_burn_amount)?;
+            stx_burn_to_add.push((principal.clone(), next_amount));
```
this clone shouldn't be necessary

kantai commented on src/vm/contexts.rs:
```diff
@@ -186,6 +240,26 @@ impl AssetMap {
+                map.insert(principal.clone(), HashMap::new());
+                map.get_mut(&principal).unwrap()
+            };
+            output_map.insert(AssetIdentifier { contract_identifier: QualifiedContractIdentifier::transient(), asset_name: ClarityName::from("STX") }, AssetMapEntry::STX(stx_amount as u128));
```
We should probably have a built-in contract identifier for STX, rather than using the transient identifier.

jcnelson replied:
How about all 0's?  That's the contract address of the chain's "boot code" (see `src/chainstate/stacks/db/mod.rs`).

kantai replied:
Sounds good to me

kantai commented on src/vm/database/marf.rs:
```diff
@@ -62,10 +62,7 @@ pub fn in_memory_marf() -> MarfedKV<SqliteConnection> {
-///   chain_tip. If the MARF already has some chain_tips, then it selects ordinal 0.
-///   Otherwise, it uses the block_sentinel.
-pub fn sqlite_marf(path_str: &str, chain_tip: Option<BlockHeaderHash>) -> Result<MarfedKV<SqliteConnection>> {
+pub fn sqlite_marf(path_str: &str, miner_tip: Option<&BlockHeaderHash>) -> Result<MarfedKV<SqliteConnection>> {
```
It's usually better for the function argument to accept an owned reference in situations like this -- that allows the caller to decided whether or not a clone() is necessary -- otherwise, the function will _always_ need to clone, even if the caller has an owned reference that it's not ever using again

kantai commented on src/vm/clarity.rs:
```diff
@@ -152,16 +248,20 @@ impl <'a> ClarityBlockConnection <'a> {
     /// An error here would indicate that something has gone terribly wrong in the processing of a contract insert.
     ///   the caller should likely abort the whole block or panic
-    pub fn save_analysis(&mut self, identifier: &QualifiedContractIdentifier, contract_analysis: &ContractAnalysis) -> Result<(), CheckError> {
+    pub fn save_analysis(&mut self, identifier: &QualifiedContractIdentifier, contract_analysis: &ContractAnalysis) -> Result<(), Error> {
```
I don't think we should reduce the specificity of the return type here -- this function _can only ever_ throw a CheckError, there's no reason for this function to return a different type of error when any caller can either use `Error::from` or `.into` or `?` to handle the conversion, whereas a caller that is expecting to handle a CheckError would need to do a spurious match with unreachable paths.

jcnelson replied:
Sure, I can fix that.  IIRC I made that change to make it easier to write the transaction-processing code -- this change is not functionally required.

kantai commented on src/vm/database/clarity_db.rs:
```diff
@@ -455,3 +466,46 @@ impl <'a> ClarityDatabase <'a> {
+// load/store STX token state and account nonces
+impl<'a> ClarityDatabase<'a> {
+    fn make_key_for_account(principal: &PrincipalData, data: StoreType) -> String {
+        format!("vm::{}::{}", principal, data as u8)
```
This should probably have a different prefix from the `vm::` one used for general contract data to more explicitly avoid any keyspace overlap -- maybe `vm-account::{}::{}` ?

jcnelson replied:
Can do

kantai commented on src/chainstate/burn/mod.rs:
```diff
@@ -24,16 +24,17 @@ pub mod distribution;
 pub mod sortition;
 
-pub const CHAINSTATE_VERSION: &'static str = "21.0.0.0";
+pub const CHAINSTATE_VERSION: &'static str = "22.0.0.0";
```
Should this be defined in `core/mod.rs` alongside the byte array version string?

jcnelson replied:
Sure, I can move it there.

kantai commented on src/chainstate/burn/mod.rs:
```diff
@@ -74,20 +79,39 @@ impl VRFSeed {
+        let h = Sha512Trunc256Sum::from_data(&proof.to_bytes());
+        let mut b = [0u8; 32];
+        b.copy_from_slice(h.as_bytes());
+        VRFSeed(b)
```
This could just be `VRFSeed(h.0)` and avoid a new allocation.

kantai commented on src/chainstate/burn/mod.rs:
```diff
@@ -236,32 +265,21 @@ impl ConsensusHash {
+    pub fn get_prev_consensus_hashes<'a>(tx: &mut BurnDBTx<'a>, block_height: u64, first_block_height: u64, tip_block_hash: &BurnchainHeaderHash) -> Result<Vec<ConsensusHash>, db_error> {
         let mut i = 0;
         let mut prev_chs = vec![];
         while i < 64 && block_height - (((1 as u64) << i) - 1) >= first_block_height {
```
I think loops like these are clearer and have less repetition if they iterate on blocks_back directly, rather than recomputing (2**i - 1), i.e.,

```
let mut cur_blocks_back = 0;
while cur_blocks_back < u64::max && block_height >= cur_blocks_back && block_height - cur_blocks_back >= first_block_height {
   let prev_block = block_height - cur_blocks_back;
   ...
   cur_blocks_back = (cur_blocks_back << 1) + 1
}
```

otherwise, the code ends up repeating the same calculation (left-shift and subtract) a bunch of times.

jcnelson replied:
Hmm, I have the opposite opinion -- I like seeing the full mathematical expression (even if it's repeated) because it makes the code easier to follow.  When evaluating how the code works in my mind, I don't have to remember how `cur_blocks_back` changes over multiple loop passes.  But, I'll happily make the change if this is the minority opinion.

kantai commented on src/burnchains/bitcoin/bits.rs:
```diff
@@ -54,107 +67,33 @@ pub fn parse_script<'a>(script: &'a Script) -> Vec<Instruction<'a>> {
     return ret;
```
`parse_script` can be replaced with a single `.collect()`:

```rust
pub fn parse_script<'a>(script: &'a Script) -> Vec<Instruction<'a>> {
    script.iter(false).collect()
}
```

jcnelson replied:
Addressed in 9f86c7a4

kantai commented on src/chainstate/burn/db/burndb.rs:
```diff
@@ -22,132 +22,95 @@ use rusqlite::types::ToSql;
-        }
+// for using BurnchainHeaderHash values as block hashes in a MARF
+impl From<BurnchainHeaderHash> for BlockHeaderHash {
+    fn from(bhh: BurnchainHeaderHash) -> BlockHeaderHash {
```
This can just be `BlockHeaderHash(bhh.0)` and avoid an allocation and copy -- same goes for `From<BlockHeaderHash> for BurnchainHeaderHash`

jcnelson replied:
Addressed in a8f13b1b.

kantai commented on src/util/db.rs:
```diff
@@ -129,10 +166,10 @@ where
     T: FromRow<T>
```
The `FromRow` and `RowOrder` traits are a dangerous way to implement our database query interactions -- there are a number of database queries with a large number of columns, the order of which determines how they get parsed into objects, and many fields have the same type -- meaning that simple bugs in the ordering and corresponding indexing would result in Errors in the best case, and silent bugs in the worst case. Using the column names would remove the necessity of the index argument (and types that only need _one_ column of the row like hashes, etc., should implement FromSQL rather than FromRow) and avoid the possibility of those kinds of bugs.

jcnelson replied:
:100:.  I didn't know you could index a row's fields by column in rusqlite, but it turns out [you can](https://docs.rs/rusqlite/0.20.0/rusqlite/struct.Statement.html).  I will make this change.

kantai replied:
Yep -- the Row::get function accepts RowIndex's which is implemented for usize and &str

jcnelson replied:
I opened issue #1188 which will track progress on this.  I don't think it would break anything to leave the database traits as-is in this PR, but I'll gladly send a separate PR to fix this.

kantai commented on src/burnchains/burnchain.rs:
```diff
@@ -305,322 +447,268 @@ impl Burnchain {
+    /// Return the ordered list of blockstack operations by vtxindex
+    fn get_blockstack_transactions(burnchain: &Burnchain, block: &BurnchainBlock, block_header: &BurnchainBlockHeader) -> Vec<BlockstackOperationType> {
+        debug!("Extract Blockstack transactions from block {} {}", block.block_height(), &block.block_hash().to_hex());
+        let mut ret = vec![];
```
Something like this can be more simply and idiomatically implemented with a [filter_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) and collect() -- you also shouldn't need to clone the blockstack_op

jcnelson replied:
Addressed in baa87c05

kantai commented on src/chainstate/stacks/address.rs:
```diff
@@ -21,24 +21,238 @@ use net::StacksMessageCodec;
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<StacksAddress, net_error> {
+        let mut index = *index_ptr;
+
+        let new_index = index.checked_add(STACKS_ADDRESS_ENCODED_SIZE).ok_or(net_error::DeserializeError)?;
```
What is `new_index` used for? Is it only to bounds check the additions further down? If so, it'd be better to do the bounds check on the final indexes directly. I think there's an off-by-one error as the checks are currently written -- `new_index` is checked for > buf.len, but it should be an overflow if its >= buf.len.

jcnelson replied:
Good catch -- that should be >=.  `new_index` is just used for the bounds-check.  However, `index` will be (implicitly) advanced to the value of `new_index` here because we use it to read a `u8` and a `Hash160`, or 21 bytes (which is the same value as `STACKS_ADDRESS_ENCODED_SIZE`.  I can make this clearer.

jcnelson replied:
Fixed in 3cbc89e1

jcnelson commented on src/chainstate/stacks/db/accounts.rs:
```diff
@@ -0,0 +1,829 @@
+    /// There must be MINER_REWARD_WINDOW items in the sample.
+    ///
+    /// TODO: .burns?
+    fn calculate_miner_reward(miner: &MinerPaymentSchedule, sample: &Vec<(MinerPaymentSchedule, Vec<MinerPaymentSchedule>)>) -> MinerReward {
```
Don't bother reading this method for now -- it is broken to the point where it will need to be rewritten.  Sorry for not catching that sooner!

jcnelson commented on src/chainstate/stacks/db/accounts.rs:
```diff
@@ -0,0 +1,829 @@
+}
+
+#[cfg(test)]
+mod test {
```
These tests need to be correct to compensate for the broken `calculate_miner_reward` method.

kantai commented on src/chainstate/burn/operations/leader_key_register.rs:
```diff
@@ -108,89 +133,89 @@ where
+    fn check<'a>(&self, burnchain: &Burnchain, block_header: &BurnchainBlockHeader, tx: &mut BurnDBTx<'a>) -> Result<(), op_error> {
+        // this will be the chain tip we're building on
+        let chain_tip = BurnDB::get_block_snapshot(tx, &block_header.parent_block_hash)
+            .expect("FATAL: failed to query parent block snapshot")
```
Should this panic? If a leader submits a bad block header with a bad parent hash, that should just get ignored by network nodes -- or is that checked previously in the block processing, and if the code reaches this point, it should panic?

jcnelson replied:
`block_header.parent_block_hash` is the Bitcoin block's parent block hash, not the Stacks block's parent block hash.  The `block_header` value comes from data fetched by `Burnchain::sync`, which pulls block headers and blocks from Bitcoin.  Bitcoin blocks are processed in hash-chain order, so a parent block's snapshot should always be calculated and stored to the `BurnDB` before any of its children.  If `BurnDB::get_block_snapshot` fails to load a given Bitcoin block's parent block's snapshot, it indicates a serious (unrecoverable) problem.

kantai commented on src/chainstate/burn/operations/mod.rs:
```diff
@@ -150,3 +260,26 @@ where
 
+
+// parser helpers
+pub fn parse_u32_from_be(bytes: &[u8]) -> Option<u32> {
```
The rust u32 and u16 primitives provide big endian parse functions that should be used -- not just because we shouldn't reimplement primitives, but also because it won't require casting the u8s to u32s for bitshifts

```rust
fn parse_u32_from_be(bytes: &[u8]) -> Option<u32> {
    x.try_into().ok().map(u32::from_be_bytes)
}
```

jcnelson replied:
Didn't know that at the time it was written.  Will fix.

jcnelson replied:
Fixed in 8f8c6e10

kantai commented on src/chainstate/stacks/auth.rs:
```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+        let fields : Vec<TransactionAuthField> = read_next(buf, &mut index, max_size)?;
+        let signatures_required: u16 = read_next(buf, &mut index, max_size)?;
+        
+        // read and decode _exactly_ num_signatures signature buffers
```
this comment doesn't appear accurate -- it looks like line 151 will parse however many TransactionAuthFields are supplied in the input

jcnelson replied:
A `TransactionAuthField` can be a public key or a signature.  The number of signatures required can be less than the number of auth fields.  Right now, there's no limit on how many auth fields a transaction can have; only that `num_signatures` must match the number of `TransactionAuthField::Signature`s present in the `fields` vector.  I expect that there will be a soft-fork default limit to the number of fields, but I'm not sure it's a good idea to force an upper bound at the protocol level (miners will figure that out by themselves).

kantai replied:
Gotcha -- that's fine by me. I think the comment led me to believe the function should be failing fast if more signatures are read than `signatures_required`, but that isn't really necessary.

kantai commented on src/chainstate/stacks/auth.rs:
```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+    /// to by order of signers (instead of re-serializing the tranasction each time).
+    /// Calculates the next sighash and public key, which the next verifier must verify.
+    /// Used by StacksTransaction::verify*
+    pub fn next_verification(cur_sighash: &Txid, cond_code: &TransactionAuthFlags, key_encoding: &TransactionPublicKeyEncoding, sig: &MessageSignature) -> Result<(StacksPublicKey, Txid), net_error> {
```
Where does this function verify the current signature? I see digest_bits being filled, but they don't appear to get checked.

jcnelson replied:
Ah, so this algorithm is kind of interesting.  The top-level verification methods you want to look at are `SinglesigSpendingCondition::verify()` and `MultisigSpendingCondition::verify()`, which check the output of this method.

The call to the `recover_to_pubkey` method works like an ECDSA signature verification in that it takes the _(r, s)_ signature point, the plaintext digest, and a 1-byte "recovery code," and turns it back into a public key (the recovery code indicates _which_ of a set of up to two recovered public keys to use).  At the very end of verifying the transaction in `SinglesigSpendingCondition::verify()` and `MultisigSpendingCondition::verify()`, the public key(s) are hashed and compared to the address of the spending condition, which completes the check.  Note that using a different plaintext digest for the same _(r, s)_ signature will produce a _different_ public key, meaning that if the plaintext is modified, then the resulting public key hash will not match the address.

kantai replied:
I see -- so the signature verification is happening in `StacksPublicKey::recover_to_pubkey` -- in that case, the `digest_bits` variable is an unused variable?

jcnelson replied:
Yup; will remove.

jcnelson replied:
Fixed in d399b6d9

kantai commented on src/chainstate/stacks/auth.rs:
```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+    }
+
+    #[test]
+    fn tx_stacks_invalid_spending_conditions() {
```
Does this test cover the `VerifyingError("Invalid Message...")`, `VerifyingError("Invalid Signature: failed to decode...")`, and `VerifyingError("Invalid Signature: failed to recover...")` cases? It'd be more clear if this test checked for error equality rather than just `is_err()`.

jcnelson replied:
No; I'll add that.

kantai commented on src/chainstate/stacks/db/accounts.rs:
```diff
@@ -0,0 +1,829 @@
+    }
+    
+    /// Verify that this transaction is not a replay, and update its nonce.
+    pub fn update_account_nonce<'a>(clarity_tx: &mut ClarityTx<'a>, tx: &StacksTransaction, account: &StacksAccount) {
```
Where is the nonce of the transaction checked?

jcnelson replied:
I don't think it is (oops).  Will fix -- it should be checked in `process_transaction()` in `src/chainstate/stacks/db/transactions.rs`.

jcnelson replied:
Fixed in ef2e7c6c

kantai commented on src/chainstate/stacks/db/accounts.rs:
```diff
@@ -0,0 +1,829 @@
+                };
+            
+            test_debug!("{}: Anchored tx fees shared = {} * ({}/{})", miner.address.to_string(), anchored_tx_fee_shared_total, num_mined, sample.len());
+            anchored_tx_fee_shared = anchored_tx_fee_shared_total.checked_mul(num_mined).expect("FATAL: STX shared anchored tx fee overflow") / (sample.len() as u128);
```
I'm not sure if this is the right ratio here -- the shared fees (according to SIP-1) should be _T_ * _P_b_ where _P_b_ is the ratio of work done by this miner over the total work of miners in that window (and _T_ is the shared reward). This code calculates _T_ * (blocks_participated_in / number_of_blocks), which is a different ratio, and also problematic because user support burns, would allow many miners to participate in a single block, meaning that the split of shares wouldn't necessarily sum to 1, e.g., in a window of 2 blocks:

block 1 = mined by Alice, supported by Bob
block 2 = mined by Bob, supported by Alice

Both Alice and Bob would have num_blocks = 2, and thus each would claim 100% share of the total shared tx fee.

kantai replied:
Oops -- missed your previous comment.

kantai commented on src/chainstate/burn/operations/leader_block_commit.rs:
```diff
@@ -135,180 +137,211 @@ where
+        }
         
         /////////////////////////////////////////////////////////////////////////////////////
         // This tx's epoch number must match the current epoch
```
This epoch check is no longer performed, right? This comment should be removed

jcnelson replied:
Correct; will remove.

jcnelson replied:
Fixed in 8f8c6e10

kantai commented on src/chainstate/stacks/db/transactions.rs:
```diff
@@ -0,0 +1,1001 @@
+
+    /// Process a token transfer payload (but wrapped in its transaction still, in order to do
+    /// post-condition checks).
+    fn process_transaction_token_transfer<'a>(clarity_tx: &mut ClarityTx<'a>, tx: &StacksTransaction, origin_account: &StacksAccount) -> Result<(), Error> {
```
Since this function panic when it's supplied anything but a TokenTransfer, it'd be simpler to type the argument `tx` as `&TransactionTokenTransfer`, that way the type checker will enforce the match, rather than a check and panic.

jcnelson replied:
Ah, I didn't know you could do that.  Will change.

Related, I think we're dropping support for non-STX token transfers?  I might just remove most of this code outright.

jcnelson replied:
Okay, so I can't actually change this -- we need to pass the whole tx into this method in order to evaluate post-conditions.

kantai replied:
Ah, yes, I see. Somewhat relatedly, why would a transfer need post-conditions?

jcnelson replied:
I guess we don't need them for STX transfers -- it's not like they add any value in this particular case.  I'm fine with just requiring that these transactions have zero post-conditions.

Speaking of things that should or should not be allowed in STX transfers, I did think of a reason why you'd want to support sponsored STX transfers (which are currently prohibited) -- I would want to do this for transferring STX from code storage.  I would sign my STX transaction with my cold storage key on an air-gapped computer, and then use a separate (much-lower-value) account to sponsor the transaction and relay it safely.  Importantly, I don't need to guess the fee when I'm on the air-gapped computer -- I can wait for an arbitrarily long amount of time before sending the cold-storage-signed transaction, since the use of a separate sponsoring key effectively gives me a way to late-bind on the fee.  I went ahead and removed this prohibition in my tree.

kantai replied:
Agreed -- I think _any_ transaction should be sponsorable (similar to a sighash anyone can pay). Allowing a separate key to pay transaction fees is useful for any number of reasons.

jcnelson replied:
Addressed in d399b6d9113660aa1a71e88d9e7915c6dd8bb12f -- the transaction sighash for origin accounts no longer commits to anything in sponsor accounts (only that the transaction will be sponsored by someone), so a sponsor can sign a transaction already signed by the origin account and commit to its own nonce and tx fee (neither of which the origin knows).  This should be equivalent to sighash-anyone-can-pay.

kantai commented on src/chainstate/stacks/auth.rs:
```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+
+impl TransactionSpendingCondition {
+    pub fn new_singlesig_p2pkh(pubkey: StacksPublicKey) -> Option<TransactionSpendingCondition> {
+        let signer_addr = match StacksAddress::from_public_keys(0, &AddressHashMode::SerializeP2PKH, 1, &vec![pubkey.clone()]) {
```
The `?` operator also works with `Option<..>` types and can be pretty helpful in lines like this

jcnelson replied:
Didn't know that; will update.

jcnelson replied:
Addressed in d399b6d9113660aa1a71e88d9e7915c6dd8bb12f

kantai commented on src/util/strings.rs:
```diff
@@ -0,0 +1,321 @@
+        let bytes = read_clarity_string_bytes(buf, &mut index, max_size)?;
+
+        // must encode a valid string
+        let s = String::from_utf8(bytes.clone())
```
The clone here and in the other deserialize function can be removed

kantai commented on src/chainstate/burn/sortition.rs:
```diff
@@ -135,30 +131,24 @@ impl BlockSnapshot {
+            .expect("FATAL: failed to query last snapshot with sortition");
+
+        let VRF_seed =
+            if last_sortition_snapshot.sortition_hash == SortitionHash::initial() {
```
Should this if condition call `is_initial()`?

jcnelson replied:
It shouldn't make a difference -- absent a hash collision, there's no way to get a sortition hash whose bytes are all 0's.  I don't think the `is_initial()` method is even used anywhere.

kantai replied:
Right -- we should either delete `is_initial()` because its unused, or use it to avoid duplication of logic

jcnelson replied:
Fixed in 8f8c6e10

kantai commented on src/chainstate/burn/sortition.rs:
```diff
@@ -173,147 +163,220 @@ impl BlockSnapshot {
+        // Try to pick a next block.
+        let (winning_txid, winning_block_hash, winning_block_vrf_seed) = match BlockSnapshot::select_winning_block(tx, block_header, &next_sortition_hash, burn_dist)? {
+            None => {
+                // should be unreachable, but would happen if there were no burns
```
Since this function checks for the 0 burn case above, should this branch panic? (or call .expects on the select_winning_block result)

jcnelson replied:
Yes -- this branch is unreachable because the only way that this method returns `None` is if `burn_dist` has length 0 (and this case is handled earlier in this method).

jcnelson replied:
Fixed in 8f8c6e10

kantai commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -0,0 +1,2427 @@
+
+        let mut end = 0;
+        let mut connects = false;
+        for i in 0..num_microblocks {
```
should this be iterating over the signed_microblocks ?

jcnelson replied:
yes; will fix

jcnelson replied:
Fixed in b540a484ee9e610559b3a318f2946db2ecb7d6ce

kantai commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -0,0 +1,2427 @@
+        for tx in block.txs.iter() {
+            let (tx_fee, tx_burns) = StacksChainState::process_transaction(clarity_tx, tx)?;
+            fees = fees.checked_add(tx_fee as u128).expect("Fee overflow");
+            burns = burns.checked_add(tx_burns as u128).exp
```
The microblocks being validated and processed here are the parent block's microblocks, right? Are they associated in the database with the child's burn_header_hash and anchored_block_hash or should they be associated with the parent's?

jcnelson replied:
Yes, they're the parent's.  They're associated in the staging microblock database with the parent's burn_header_hash and anchored_block_hash.  This isn't directly stated in the docstring for `store_staging_microblock`, but it is mentioned in `preprocess_streamed_microblock` (which validates and queues up microblocks).

I have some (uncommitted) code that shows how this works in `miner.rs` (the version in Github doesn't process microblocks yet).

kantai replied:
Gotcha -- it looks like at lines 1683 and 1704, it's marking the microblocks for `next_staging_block` (the child) as processed, rather than the parent, which is what it's validating.

Also, not sure why this github comment thread is appended to line 1374 (it should be line 1683, and appears so in other views -- github bug maybe?)

jcnelson replied:
> Gotcha -- it looks like at lines 1683 and 1704, it's marking the microblocks for next_staging_block (the child) as processed, rather than the parent, which is what it's validating.

Yes; that's a bug (fixed in my tree).

jcnelson replied:
Fixed in b540a484ee9e610559b3a318f2946db2ecb7d6ce

lgalabru commented on src/burnchains/bitcoin/mod.rs:
```diff
@@ -162,3 +171,51 @@ pub enum BitcoinNetworkType {
+            block_height: height,
+            block_hash: hash.clone(),
+            parent_block_hash: parent.clone(),
+            txs: txs.clone(),
```
I'm confused by this function and the way it's being used:
Arguments `hash` and `parent` are instances of `BurnchainHeaderHash`, and hydrating `block_hash` and `parent_block_hash` (header_hash -> block_hash).

jcnelson replied:
In most blockchains, the hash of the block header and the hash of the block itself refer to the same hash, so the code here uses "header hash" and "block hash" interchangeably.

kantai commented on src/chainstate/stacks/mod.rs:
```diff
@@ -20,69 +20,192 @@
+            Error::BlockTooBigError => "Too much data in block",
+            Error::MicroblockStreamTooLongError => "Too many microblocks in stream",
+            Error::ClarityError(ref e) => e.description(),
+            Error::ClarityInterpreterError(ref e) => "Clarity fucked up and Aaron didn't make this error struct implement the Error trait, so who knows what went wrong?",
```
👀

jcnelson replied:
Lol that's not going to make it to `master`!

More seriously, I think this right here was also the reason I modified the code in `src/vm/clarity.rs` to use `Error` instead of `CheckError` (which you commented on earlier).

kantai replied:
lol, but both implement `err::Error`!

I think there's just not much useful information provided by that trait (e.g., you don't get anything at all like a stack trace, and .source is only useful for wrapped errors) -- check errors however _do_ try to provide more information via the `DiagnosableError` trait, and the interpreter runtime errors also try to provide Clarity stack information.

jcnelson replied:
should be fixed now -- this error is removed

kantai commented on src/util/strings.rs:
```diff
@@ -0,0 +1,321 @@
+
+    pub fn try_as_clarity_literal(&self) -> Option<Value> {
+        // must parse to a single Clarity literal
+        match lex(&self.to_string()) {
```
Clarity values should be serialized / deserialized as JSON, rather than lexed as Clarity code -- if you use `Value::try_deserialize()`, I think that should do the trick.

https://github.com/blockstack/blockstack-core/blob/develop/src/vm/types/serialization.rs#L346

jcnelson replied:
This method get used in the `TransactionPayload::ContractCall` variant to parse a `contract-call?`'s arguments, which must be literals (it's also used to represent NFTs in the token-transfer transaction payload, but that is slated to be dropped).  This means whatever serialization we use here will be set in stone once the system goes live.  Do we intend to keep using JSON, or is a more economical wire format planned?

kantai replied:
My understanding was that we had decided to use JSON. I am certainly open to using a different serialization format --- either way, Values should be serialized and deserialized using that Value serialization function.

jcnelson replied:
Ah.  My understanding was that we were using JSON as a stop-gap, but we would use something more economical (and that offers an unambiguous serialization) for the release.  I'll switch to using the `serialize` and `deserialize` routines here either way.

kantai replied:
Sounds good -- we can revisit it, but as it stands in SIP-004, Clarity values are serialized as JSON, and I think there's probably good reason to revisit it due to bandwidth concerns.

jcnelson replied:
For now, I just ended up implementing `StacksMessageCodec` for `Value`, which internally uses `ClaritySerializable::serialize` for turning a `Value` into bytes and back.  The `TransactionPostCondition::Nonfungible` and `TransactionPayload::ContractCall` variants have been updated to use `Value` instead of `StacksString` for representing actual Clarity values, thereby obviating this method altogether.  The `is_clarity_variable()` method in this struct is still needed for determining that the function name in `TransactionPayload::ContractCall` is still well-formed, however.

jcnelson replied:
Fixed in 4f34dedd2c8b1ee4475f79c5fd9405e66cf63151.

kantai commented on src/util/vrf.rs:
```diff
@@ -192,214 +380,202 @@ impl ECVRF_Proof {
+    }
 }
 
 #[cfg(test)]
```
Should this VRF implementation be able to use the test vectors supplied in https://tools.ietf.org/id/draft-irtf-cfrg-vrf-04.html#rfc.appendix.A?

jcnelson replied:
If any apply, it would be those in A.4 -- specifically, the Ed25519-SHA512-Elligator2 test vectors.  However, I'm not sure they apply at all, since our VRF doesn't use the Elligator2 map (but uses the Elligator map on the Ristretto subgroup, defined here: https://ristretto.group/details/elligator.html).  In fact, our hash-to-curve routine (line 398) does not currently use the RFC-recommended hash-to-curve routine for Ed25519-SHA512-Elligator2 (defined here: https://tools.ietf.org/id/draft-irtf-cfrg-vrf-02.html#suites) to avoid confusion.

kantai replied:
According to curve25519_dalek -- Elligator2 is the map used (see https://doc.dalek.rs/curve25519_dalek/ristretto/struct.RistrettoPoint.html#method.hash_from_bytes). Is there a reason to implement our own hash-to-curve routine? That `hash_from_bytes` function implements hash-to-curve for a given digest.

jcnelson replied:
I stand corrected!  I thought they were different Elligator maps.

> Is there a reason to implement our own hash-to-curve routine?

By this, I mean we use a different suite version number (which gets hashed with the public key and alpha).  I'll try it out with the suite version recommended by the RFC.  If we can use the RFC test vectors, we definitely should :+1:

jcnelson replied:
Fun story -- the Ristretto algorithm for converting a hash to a point is _not_ the same process for converting a hash to an Ed25519 point.  The Ristretto algorithm applies the Elligator map _twice_ -- once on the upper 32 bytes of the hash, and once on the lower 32 bytes of the hash -- and adds the result.  This is not described by the RFC.

Interestingly, Libra's VRF _does_ follow the RFC, and even uses the same libraries we do: https://github.com/libra/libra/blob/master/crypto/crypto/src/vrf/ecvrf.rs

jcnelson replied:
If RFC compliance is the goal here, an alternative is to use a different hash-to-curve algorithm like the try-and-increment algorithm here: https://tools.ietf.org/id/draft-irtf-cfrg-vrf-02.html#ecvrfH2C1.  It's not a constant-time operation, but that's not necessary for us because the alpha parameter (the VRF seed) is public, not secret.

Coincidentally, this is the hash-to-curve algorithm Libra uses in its VRF (not Elligator2).

jcnelson replied:
Okay, I re-implemented the VRF to be RFC-compliant.  The ECVRF-ED25519-SHA512-TAI test vectors all pass (we're now using the try-and-increment hash-to-point scheme instead of the elligator map).  Also, all operations are now on full-bore Ed25519 instead of Ristretto points.

kantai replied:
Awesome!

jcnelson replied:
Addressed in a2320797

lgalabru commented on src/chainstate/burn/operations/mod.rs:
```diff
@@ -61,86 +106,151 @@ impl error::Error for Error {
+    pub parent_vtxindex: u16,               // offset in the parent block where the parent block hash can be found
+    pub key_block_ptr: u32,                 // pointer to the block that contains the leader key registration 
+    pub key_vtxindex: u16,                  // offset in the block where the leader key can be found
+    pub memo: Vec<u8>,                      // extra unused byte
```
I'm confused by the typing here: should we have 
```
    pub parent_block_ptr: u64,
    pub parent_vtxindex: u32,
    pub key_block_ptr: u64,
    pub key_vtxindex: u32,
```
instead?

jcnelson replied:
In this case, the sizes of these fields correspond to the sizes in the OP_RETURN.

lgalabru commented on src/chainstate/burn/operations/mod.rs:
```diff
@@ -61,86 +106,151 @@ impl error::Error for Error {
+    pub txid: Txid,                         // transaction ID
+    pub vtxindex: u32,                      // index in the block where this tx occurs
+    pub block_height: u64,                  // block height at which this tx occurs
+    pub burn_header_hash: BurnchainHeaderHash,      // hash of the burn chain block header
```
Keep being confused by the naming + comment:
`hash of the burn chain block header`, or `hash of burn chain block` (same file, line 177) ?

Either way, this field is supposed to be hydrated with the hash of the parent block, not the hash of the block including this op, right?
If so, there might be an issue in `sortition.rs`:
https://github.com/blockstack/blockstack-core/blob/8635433074d30e7567e8578ebf6f5303b08716cf/src/chainstate/burn/sortition.rs#L148-L150

Where I think we should be fetching the `op` using the parent block hash of the block where the sortition is happening, instead of `last_sortition_snapshot.burn_header_hash`.

lgalabru commented on src/chainstate/stacks/db/contracts.rs:
```diff
@@ -0,0 +1,127 @@
+use std::io::prelude::*;
+use std::fmt;
+use std::fs;
+use hashbrown::HashMap;
```
If we're not using`hashbrown::HashSet`, we can get rid of this `hashbrown` dependency, it became the default implementation of `HashMap` with Rust 1.36 - https://github.com/rust-lang/hashbrown

jcnelson replied:
Ah, didn't know that.  Sounds good to me :+1:

jcnelson replied:
Fixed as of now.

lgalabru commented on src/chainstate/stacks/address.rs:
```diff
@@ -21,24 +21,238 @@ use net::StacksMessageCodec;
+}
+
+impl Address for StacksAddress {
+    fn to_bytes(&self) -> Vec<u8> {
```
💅 should we stick to rust conventions, and name this method `into_bytes`, instead of `to_bytes`?

jcnelson replied:
I think `to_bytes()` means "create a byte representation without destroying the object" whereas `into_bytes()` means "convert the object into its byte representation".  The semantics of `to_bytes()` are desired here.

jcnelson commented:
I think this last flurry of commits addresses all of the outstanding comments except for the broken fee calculation, which I will fix this week.  It won't be complete, however, until the runtime cost calculation code is merged in.

kantai commented:
>  It won't be complete, however, until the runtime cost calculation code is merged in.

This functionality should just be shimmed for the time being -- the components in this PR should be testable independent of the runtime cost calculation code.

jcnelson commented on src/burnchains/mod.rs:
```diff
@@ -219,9 +342,791 @@ impl error::Error for Error {
+            }
+        }
+
+        pub fn sign_as_sposnor(&self, tx_signer: &mut StacksTransactionSigner) -> () {
```
Fixed

jcnelson commented:
Test coverage that is still missing and/or would fail:

Burnchain tests:
* VRF leader keys should expire after some amount of time (exact time TBD)

MARF tests (related to block-processing):
* Writing a Trie file is atomic, but not crash-consistent.  There should be a test that the MARF should be able to identify corrupt or missing Tries, so that the block-processing loop can re-generate them from blocks on disk.

Account processing:
* C32 addresses should be normalized before they are decoded (we currently don't test for this)

Transaction processing:
* Should be able to receive STX out of the miner reward contract in the boot code
   * Blocked on the absence of `stx-transfer!`
* Should be able to test miner reward maturation even if we can't transfer STX yet (there is commented-out code for this, but it's broken).
* Should be able to test calls into the miner reward contract to see if STX are ready to be transferred out.
* Should be able to withdraw no more STX from the miner reward contract than are currently due to the miner.
* Should only be able to withdraw from the miner reward contract if the miner is authorized.
* Should be able to ensure that a transactions is not processed if its post-condition mode is set to `Deny` and an unaccounted-for asset got moved.
* Should be able to generate and process poison-microblock transactions, which slash the volating miner's coinbase and award 5% of it to the slasher.  This would entail testing that the transaction altered the record for the miner in the miner reward contract before the maturity period ends, as well as testing that a transaction like this would be invalid if processed after the reward in question was matured.

Block processing:
* Should be able to see the block reward go up in response to burning STX
   * Blocked on the absence of `stx-burn!`
* Should calculate STX fee rate as K+1'th highest fee rate.
* Should be able to stub out a way to get a transaction's total cost, so we can determine how much to bill a transaction's paying account.
* Should be able to measure how "full" an epoch is, even with a cost stub and stubs for the compute budgets of the epoch.
* Should be able to measure how big blocks and microblocks are, and not mine blocks that are too big, even if we don't yet know what the bandwidth budget is.

Mining:
* Make sure that our checks for mining invalid blocks are actually working
* Make sure we can mine blocks even if we go for multiple Bitcoin blocks without a sortition
* Make sure we can mine multiple anchored blocks that point to different points in the parent microblock stream.
* Make sure that we can get MARF proofs off of state-transitions that occur in the same fork as the block we're going to mine.

lgalabru commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -0,0 +1,3495 @@
+
+impl StagingBlock {
+    pub fn is_genesis(&self) -> bool {
+        self.parent_burn_header_hash == FIRST_BURNCHAIN_BLOCK_HASH && self.parent_anchored_block_hash == FIRST_STACKS_BLOCK_HASH
```
While working on #1202, I noticed this line. I have a doubt concerning the 
`self.parent_burn_header_hash == FIRST_BURNCHAIN_BLOCK_HASH` part of the test. 
If my understanding is correct, this might be an issue in a real world situation, since *parent_burn_header_hash* of the genesis block will not be the genesis block of bitcoin.. correct?

jcnelson replied:
Thanks!  Removed that check; will see if the tests continue to pass.

lgalabru replied:
Thanks! Looking at the latest commits, I think there's a similar issue here:
https://github.com/blockstack/blockstack-core/blob/b0b0fb7199929f957f09ef202b74822881d1f0c2/src/chainstate/stacks/db/blocks.rs#L1780

jcnelson replied:
Yup, will remove that too (and other instances in the codebase).

jcnelson commented:
Okay, I think I've addressed all the remaining concerns:
* Test coverage is just over 83% now
* Block reward logic is stubbed out, but we accurately test coinbases
* The block database accurately reassembles the chainstate even if blocks and microblocks arrive in random order, no matter how many forks (Stacks or BTC) exist.

kantai commented on src/vm/contexts.rs:
```diff
@@ -201,6 +290,74 @@ impl AssetMap {
+        }
+    }
+
+    pub fn get_fungible_token_ids(&self, principal: &PrincipalData) -> Vec<AssetIdentifier> {
```
This and the next function don't appear to be used anywhere:

```
$ find src -name "*.rs" -exec grep "get_fungible_token_ids" {} /dev/null \;
src/vm/contexts.rs:    pub fn get_fungible_token_ids(&self, principal: &PrincipalData) -> Vec<AssetIdentifier> {
$ find src -name "*.rs" -exec grep "get_nonfungible_token_ids" {} /dev/null \;
src/vm/contexts.rs:    pub fn get_nonfungible_token_ids(&self, principal: &PrincipalData) -> Vec<AssetIdentifier> {
```

And so should be removed.

jcnelson replied:
Done

kantai commented on src/chainstate/stacks/transaction.rs:
```diff
@@ -17,137 +17,3215 @@
+        }
+    }
+
+    pub fn get_incomplete_tx(&self) -> StacksTransaction {
```
This is a duplicate of `get_tx_incomplete`

jcnelson replied:
Lol how'd that happen?  Removed.

lgalabru commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -0,0 +1,3459 @@
+    /// return the index in the given microblocks vec that corresponds to the highest valid
+    /// block -- i.e. the microblock indicated by the anchored header as the parent.
+    /// If there was a duplicate sequence number, then also return a poison-microblock
+    /// transacti
```
Should we use `BurnDB::get_last_snapshot_with_sortition` instead of ` BurnDB::get_block_snapshot_in_fork`?
ie can we end up in a situation where the block at height (block_height -1) does not include a sortition but the block at height (block_height - 2) does?

jcnelson replied:
I'm not sure where you're referring to?  This method doesn't call either of those.  Do you mean in `validate_anchored_block_burnchain()`?

lgalabru replied:
Oh that is weird bug from github...

<img width="884" alt="Screen Shot 2020-01-03 at 10 23 48 AM" src="https://user-images.githubusercontent.com/87777/71731638-41b72780-2e13-11ea-881c-9c16bd3d0c24.png">

jcnelson replied:
Ah, thanks.  The reason we use `get_last_snapshot_in_fork` instead of `get_last_snapshot_with_sortition` is because we're interested in getting the latest `sortition_hash` field, even if no sortition happened in the last Bitcoin block.  Miners always prove on the latest `sortition_hash`, and the sortition code ensures that it will be both unique to each Bitcoin block (even if there is no sortition) and unpredictable calculating it from the hash of the last Bitcoin block's sortition hash, the winning VRF seed (if a sortition happened), and the new Bitcoin block header.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
